### ISSUE TITLE
gedis client improvements

### ISSUE THREAD

**Issue by despiegk** [threefoldtech/jumpscaleX_core#695]
the fact we pull the schema's and the actor metadata over the wire for every connection we make will not work longer time because of following issues

- there can be version conflicts
- when X 3bots talk to Y 3bots this gets very heavy, on your client you'll endup with a lot of metadata & code generation

### suggested fix

- each package gets a strict version nr 
- the package is owned 100% by 1 3bot, this 3bot is registered in phonebook
- each package get's a metadata list which describes every file (relative path + hash)
- this metadata list is signed by private key of the author 3bot
- each author 3bot hosts an rsync server on top of their packages
- if a schema gets an update which is breaking, then the update method needs to be implemented to deal with it and version nr changes on the package
- when making connection over gedis the version nr per package is given back
- the client checks the version nr, if the client is behind the client will use rsync to download the new package code and verify the files based on the metadata & signature and reload the actor clients and reload the schema's which are always unique per package anyhow
- if the client sees that his version is newer than the server then he tells the server to update
- if the server gets an update request, will download the package (rsync) from the right source, check the validity and reload the actors & schemas & do the update of data for schema's
- this means metadata for schema's and actors is not going over gedis connection!
- the metadata of actors & schema's need to be stored in json format in the package
    - this metadata can be asked to any source 3bot (author 3bot)
    - this allows other languages to generate source code for clients to connect to gedis or deal with our schema's

### advantages

- clients are much faster & efficient
- better support for updates of schema's
- less mem because less metadata

### disadvantages

- all 3botservers will have to be on newest version, otherwise this will become impossible for clients to make connections to multiple servers, which means servers will have to autoupdate, to make this secure we have to introduce the signing of the packages.

**Comment by zaibon** [comment]
It seems we are trying to solve multiple problem with this proposal.
1. Versioning of the package
2. Try to decrease the amount of "metadata" data transfer between 3bot servers
3. Add a way to migrate package between breaking versions
4. Link a package to a single 3bot author using signature.

First thing I need to mention is about the way we "generate" client code at runtime. While this is nice to have when people are interacting with the package manually using kosmos, it doesn't give any benefits for a real world application. Reason being that if the package change, chances are the client using this package will have to update his application code using the client too in order to use the new package version. So the only thing we do is just always breaking the client application side cause he has no control over when he update the client code.
To me package should indeed be versioned, but client should be generated or just distributed the normal way through code repository. So users of these package can do a proper dependency management.

Now, if we still want to go that way. This proposal regarding the distribution of the metadata of the package allow the client to only pull/generate only the **first time** he connects to a new actor/package. This is because now that the version number is exchange first, the client can know and verify if he has already the generated code locally or not. This means there is no need to download the full metadata and generate code for every connection anymore.
If that is true, the all rsync sytem is not needed anymore. The metadata can still go over gedis, but it will only once per package version. All this code is already there so this is a way shorter path to the goal.

The only thing left to implement is then just the versioning and a flow to negotiate upgrade/migration.
